src/Data/Vec/Properties.agda

```diff
@@ -70,66 +70,166 @@ module _ {a} {A : Set a} where
   lookup⇒[]= : ∀ {n} (i : Fin n) {x : A} xs →
                lookup i xs ≡ x → xs [ i ]= x
   lookup⇒[]= zero    (_ ∷ _)  refl = here
-  lookup⇒[]= (suc i) (_  ∷ xs) p    = there (lookup⇒[]= i xs p)
+  lookup⇒[]= (suc i) (_ ∷ xs) p    = there (lookup⇒[]= i xs p)
 
   []=↔lookup : ∀ {n i} {x} {xs : Vec A n} →
                xs [ i ]= x ↔ lookup i xs ≡ x
   []=↔lookup = inverse []=⇒lookup (lookup⇒[]= _ _)
      (λ _ → []=-irrelevance _ _) (λ _ → P.≡-irrelevance _ _)
 
+------------------------------------------------------------------------
+-- updateAt (_[_]%=_)
+
+module _ {a} {A : Set a} where
+
+  -- Defining properties of updateAt:
+
+  -- (+) updateAt i actually updates the element at index i.
+
+  updateAt-updates : ∀ {n} (i : Fin n) {f : A → A} (xs : Vec A n) {x : A}
+    → xs              [ i ]= x
+    → updateAt i f xs [ i ]= f x
+  updateAt-updates zero    (x ∷ xs) here        = here
+  updateAt-updates (suc i) (x ∷ xs) (there loc) = there (updateAt-updates i xs loc)
+
+  -- (-) updateAt i does not touch the elements at other indices.
+
+  updateAt-minimal : ∀ {n} (i j : Fin n) {f : A → A} {x : A} (xs : Vec A n)
+    → i ≢ j
+    → xs              [ i ]= x
+    → updateAt j f xs [ i ]= x
+  updateAt-minimal zero    zero    (x ∷ xs) 0≢0 here        = ⊥-elim (0≢0 refl)
+  updateAt-minimal zero    (suc j) (x ∷ xs) _   here        = here
+  updateAt-minimal (suc i) zero    (x ∷ xs) _   (there loc) = there loc
+  updateAt-minimal (suc i) (suc j) (x ∷ xs) i≢j (there loc) =
+    there (updateAt-minimal i j xs (i≢j ∘ P.cong suc) loc)
+
+  -- The other properties are consequences of (+) and (-).
+  -- We spell the most natural properties out.
+  -- Direct inductive proofs are in most cases easier than just using
+  -- the defining properties.
+
+  -- updateAt i  is a morphism from the monoid of endofunctions A → A
+  -- to the monoid of endofunctions Vec A n → Vec A n
+
+  -- 1a. relative identity:  f = id ↾ (lookup i xs)
+  --                implies  updateAt i f = id ↾ xs
+
+  updateAt-id-relative : ∀ {n} (i : Fin n) (xs : Vec A n) {f : A → A}
+    → f ≡ id ↾¹ lookup i xs
+    → updateAt i f xs ≡ xs
+  updateAt-id-relative zero    (x ∷ xs) eq = P.cong (_∷ xs) eq
+  updateAt-id-relative (suc i) (x ∷ xs) eq = P.cong (x ∷_) (updateAt-id-relative i xs eq)
+
+  -- 1b. identity:  updateAt i id ≗ id
+
+  updateAt-id : ∀ {n} (i : Fin n) (xs : Vec A n) →
+    updateAt i id xs ≡ xs
+  updateAt-id i xs = updateAt-id-relative i xs refl
+
+  -- 2. composition:  updateAt i f ∘ updateAt i g ≗ updateAt i (f ∘ g)
+
+  updateAt-compose : ∀ {n} (i : Fin n) {f g : A → A} →
+    updateAt i f ∘ updateAt i g ≗ updateAt i (f ∘ g)
+  updateAt-compose zero    (x ∷ xs) = refl
+  updateAt-compose (suc i) (x ∷ xs) = P.cong (x ∷_) (updateAt-compose i xs)
+
```

Comment:
I suppose we could also have

```agda
updateAt-compose-relative : ∀ {n} (i : Fin n) {f g h : A → A} →
     (f ∘ g) ≡ h ↾¹ lookup i xs →
￼     updateAt i f ∘ updateAt i g ≗ updateAt i h
```

Reply:
Yes, maybe, even though it is a simple two-step inference:
```
   updateAt i f ∘ updateAt i g ≗  { updateAt-compose }
   updateAt i (f ∘ g) ≗           { updateAt-cong-relative }
   updateAt i h
```